### PR TITLE
Support for non-contiguous hartids

### PR DESCRIPTION
**Pull request by p12nGH-zz**

### CODE REVIEW

File: spike_main/spike.cc [hunk 1]
```diff
@@ -116,7 +130,7 @@ int main(int argc, char** argv)
   if (mems.empty())
     mems = make_mems("2048");
 
-  sim_t s(isa, nprocs, halted, start_pc, mems, htif_args);
```
Comment (palmer-dabbelt) [inline spike_main/spike.cc:119]: I wouldn't bother with `std::move` here.  If you just make it a `const vector&` you can elide the copy, but it probably doesn't matter.

**Review by palmer-dabbelt — APPROVED** [review]
LGTM

**Comment by aswaterman** [comment]
There's at least one other place in Spike that assumes hart ids are contiguous: sim_t::make_dtb, where it prints cpu@ and _intc.

**Comment by p12nGH-zz** [comment]
Also some commands that take processor id as a parameter don't work correctly as well

**Comment by mwachs5** [comment]
You will probably need to change something in the debug module in here as well, as it assumes (I think) that the mapping is 1:1

**Comment by mwachs5** [comment]
https://github.com/riscv/riscv-isa-sim/blob/58b6c187df1a5cc066c97a63483d297fd159f241/riscv/debug_module.cc#L149

Here, the core writes its `mhartid` to certain addresses, and the debug module uses that value to update internal state, and return `flags` read data. to index directly into "flags". This is basically assuming a 1:1 mapping between hartsel and mhartid, which I think you are changing in this PR?

**Review by mwachs5 — CHANGES_REQUESTED** [review]
This needs some additional changes in the debug_module.cc

**Review by mwachs5 — APPROVED** [review]
I rescind my change request, since it doesn't break anything that doesn't use this flag, and doesn't use the debug module

**Comment by aswaterman** [comment]
Merging now since this won't break anything that doesn't use the new flag, and under the promise that @p12nGH will fix the devicetree and debug module soon.

**Pull request merged**